libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/usecase/UseCase.kt

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/*
+ * CLONED FROM:
+ *   https://github.com/google/iosched/blob/main/shared/src/main/java/com/google/samples/apps/iosched/shared/domain/CoroutineUseCase.kt
+ * CHANGES:
+ *  - updated package
+ *  - updated comments
+ *  - updated filename
+ *  - replaced `com.google.samples.apps.iosched.shared.result.Result` with `kotlin.Result`
+ *  - replaced `timber.log.Timber` with `com.mapbox.navigation.utils.internal.logD`
+ */
+
+package com.mapbox.navigation.dropin.usecase
+
+import com.mapbox.base.common.logger.model.Message
+import com.mapbox.base.common.logger.model.Tag
+import com.mapbox.navigation.utils.internal.logD
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.withContext
+
+/**
+ * Executes business logic synchronously or asynchronously using Coroutines.
+ */
+abstract class UseCase<in P, R>(private val coroutineDispatcher: CoroutineDispatcher) {
```

Comment:
It may be good to introduce a "single shot". But we wouldn't want a public api for exposing a generic use case like this inside the `MapboxNavigation` domain. The iosched samples repository is a domain specific SDK, so it makes sense there.

What do you think about adding this. It could work along with `MapboxNavigationObserver`
``` kotlin
object MapboxNavigationApp {
...
  fun invoke(single: MapboxNavigationSingle, coroutineDispatcher: CoroutineDispatcher)
```

``` kotlin
interface MapboxNavigationSingle {
  suspend fun invoke(mapboxNavigation: MapboxNavigation)
}
```

Reply:
All UseCases (including abstract base classes) will be internal and won't be exposed to the public.
I am not worried about the visibility just yet. We will address it later when we are closer to the release.

`MapboxNavigationSingle` is an interesting approach however I don't see a reason why we would need this. When `MapboxNavigation` changes, we can just re-instantiate UseCase instance. 
Take a look at `Provider<StopActiveGuidanceUseCase>` injection here: https://github.com/mapbox/mapbox-navigation-android/pull/5517/files#diff-b527260b6e65ce49735cfb9ee33bd4951a61723b211330761ebdc71db984c9d7
Notice that `stopActiveGuidanceUseCaseProvider.get().invoke(Unit)` will reinstantiate StopActiveGuidanceUseCase with currently attached MapboxNavigation instance before invoking it.

Reply:
Will you make the file `internal` then?

The `MapboxNavigationObserver` is public, and eventually some of the dropin ui "components" will be public. At least that was my intention. That way, developers can re-use some of the things we build.

If we intend to always keep `UseCase` internal, then is the intention to keep the philosophy for them we keep very specific to `NavigationView`?

Reply:
Yes. I will make it internal.
It's to early to say if we ever make them public. I will leave it up to leads to decide.